engine/utils/HeapArray.h

```diff
@@ -0,0 +1,735 @@
+     * @param newCount the number of objects in the array.
+     * @param newBytes the number of bytes to allocate for tracking state.
+     * @param masks a raw pointer to a set of 8-bit unsigned integers.
+     * @param data a raw pointer to the data stored in the array.
+     */
+    HeapArray(sizet& newSize, sizet& newCount, sizet& newBytes, uint8_t* masks, T* data) :
+        size {newSize},
+        count {newCount},
+        stateMasks {masks},
+        data {data}
+    {}
+
+    /**
+     * @brief A constructor for creating a HeapArray with a number of sizes.
+     * @param newSize the size of the array.
+     * @param newCount the number of elements in the array.
+     * @param masksSize the number of bytes to allocate for storing our states.
+     */
+    HeapArray(sizet& newSize, sizet& newCount, sizet& masksSize) :
+        HeapArray(newSize,
+                  newCount,
+                  masksSize,
+                  Allocate<uint8_t>(BYTE_MASK_SIZE * masksSize),
+                  Allocate<T>(sizeof(T) * newSize))
+    {}
+
+    // Functions
+
+    /**
+     * @brief Sets an element in the array to a corresponding value.
+     * @param index the index to insert the element into.
+     * @param value the element to insert.
+     */
+    void Set(const size_t& index, const T& value)
+    {
+        data[index] = value;
+    }
+
+    /**
+     * @brief Verifies the index is active. If it is not active, increment our count and set it to
+     * active.
+     * @param index the index to search for.
+     */
+    void VerifyIndex(const size_t& index)
+    {
+        CheckIfIndexInBounds(index);
+        count += !Active(index);
+        stateMasks[index / BYTE_SIZE_IN_BITS] |= (1 << (index % BYTE_SIZE_IN_BITS));
+    }
+
+    /**
+     * @brief Checks if the index is within bounds of the array. If it is not, an assertion failure
+     * will be triggered.
+     * @param index the index to evaluate.
+     */
+    void CheckIfIndexInBounds(const size_t& index)
+    {
+        assert(IsInBounds(index) && "The index provided is larger than the size of the array!");
+    }
+
+    /**
+     * @brief Checks if an array element is active. If it is not active, an assertion failure will
+     * be triggered.
+     * @param index the index to evaluate.
+     */
+    void CheckIfActive(const size_t& index)
+    {
+        assert(Active(index) && "Element does not exist in the array!");
+    }
+
+    size_t CalculateByteMaskCount(const size_t& arraySize)
+    {
+        return ((arraySize / BYTE_SIZE_IN_BITS) + 1);
+    }
+
+    /**
+     * @brief Allocates our raw pointers to the specified values.
+     * @param arraySize the size to allocate `data` to.
+     * @param bytesCount the number of bytes to assign for states.
+     */
+    void AllocateMemory(const size_t& arraySize, const size_t& bytesCount)
+    {
+        data = Allocate<T>(sizeof(T) * arraySize);
+        stateMasks = Allocate<uint8_t>(BYTE_MASK_SIZE * bytesCount);
```

Comment:
Possible memory leak

Reply:
Had a look into this. AllocateMemory is being used in two locations:

```cpp
HeapArray(const HeapArray<T>& other) : size {other.size}, count {other.count}
    {
        if (other.data == nullptr && other.size == 0) return;

        auto stateMaskCount = CalculateByteMaskCount(other.size);

        AllocateMemory(size, stateMaskCount);

        CopyData(other.data, sizeof(T) * size);
        CopyMasks(other.stateMaskBitfield, BYTE_MASK_SIZE * stateMaskCount);

        count = other.Count();
    }
```

and 

```cpp
    void Copy(const HeapArray<T>& other)
    {
        if (other.size == 0 && other.data == nullptr) return;

        count = other.count;
        size = other.size;

        auto byteCount = CalculateByteMaskCount(other.size);

        AllocateMemory(other.size, byteCount);

        CopyData(other.data, sizeof(T) * size);
        CopyMasks(other.stateMaskBitfield, BYTE_MASK_SIZE * byteCount);
    }
```

In the first case, the array will never be pre-allocated since its a copy constructor. There's no danger of a memory leak in that scenario.

In the second case, there is a bit of a risk, since the array could have some pre-allocated data. I've added a check which will de-allocate the data accordingly: 

```cpp
// If the array is currently being used then de-allocate the array and reallocate it.
        if (data != nullptr) Destroy();

        AllocateMemory(other.size, byteCount);
```

Reply:
Correction - I've changed it to reallocate memory instead. That way we can allocate memory if the array hasn't been allocated and resize it if it's a different size.